### PR TITLE
Make mlflow.pyfunc.spark_udf driver side raise exception contains conda env create command stdout/stderr output.

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes are proposed in this pull request?

Make mlflow.pyfunc.spark_udf driver side raise exception contains conda env create command stdout/stderr output
This make it easier for user to debug when conda env creation failed in UDF side.

## How is this patch tested?

https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/478391771932729/command/2674920562632835

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [x] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pyfunc/backend.py
Comment (dbczumar) [inline mlflow/pyfunc/backend.py:238]: Can we clarify what happens when `stream_output` is `True` and what happens when it's `False`? It's not very obvious from reading the code right now. Specifically, are we relying on the fact that passing `stream_output=False` to `mlflow.utils.process.exec_cmd()` captures and returns the stdout / stderr streams? If so, the naming of the variable is also pretty confusing, since both modes are streaming output; the only difference is the destination streams: sys.stderr / sys.stdout vs subprocess.PIPEs.

File: mlflow/pyfunc/__init__.py
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py:966]: Are we relying on the fact that passing `stream_output=False` to `mlflow.utils.process.exec_cmd()` captures and returns the stdout / stderr streams? I don't see those streams being used anywhere.

The naming is a bit confusing. `stream_output=False` would seem to imply that the command output shouldn't be sent anywhere at all.

Comment (harupy) [reply mlflow/pyfunc/__init__.py:966]: `capture_output` might make more sense.

**Review by dbczumar — APPROVED** [review]
LGTM once we change to `capture_output` and replace `%s` with `"".format(...)`

File: mlflow/pyfunc/backend.py
Comment (harupy) [inline mlflow/pyfunc/backend.py:228]:
```suggestion
    capture_output_while_creating_env=False,
```

Can we rename this argument?

Comment (harupy) [reply mlflow/pyfunc/backend.py:228]:
```suggestion
    capture_output_during_env_creation=False,
```

File: mlflow/pyfunc/backend.py
Comment (harupy) [inline mlflow/pyfunc/backend.py:40]: On second thought, can we split `_execute_in_conda_env` into two functions, one that creates an env and returns the env name, one that takes an env name and a command to run?

File: mlflow/pyfunc/backend.py
Comment (WeichenXu123) [reply mlflow/pyfunc/backend.py:228]: what about `capture_conda_failure_output_and_raise` which is more meaningful ?

File: mlflow/pyfunc/backend.py
Comment (WeichenXu123) [reply mlflow/pyfunc/backend.py:40]: We already have the function `get_or_create_conda_env`
but "prepare_env" still calls `_execute_in_conda_env` I think it is for running `python -c ""` immediately after conda env created in order to validate the env works ?

File: mlflow/pyfunc/backend.py
Comment (WeichenXu123) [reply mlflow/pyfunc/backend.py:228]: what about `capture_conda_failure_output_and_raise` which is more meaningful here ?

File: mlflow/utils/conda.py
Comment (harupy) [inline mlflow/utils/conda.py:190]: Can we print out the exception using `_logger.warning`?

File: mlflow/pyfunc/backend.py
Comment (harupy) [reply mlflow/pyfunc/backend.py:40]: Can we pull `get_or_create_conda_env` out of `_execute_in_conda_env`?

File: mlflow/pyfunc/backend.py
Comment (harupy) [reply mlflow/pyfunc/backend.py]: I think `capture_conda_failure_output_and_raise` is very confusing and difficult to guess its behavior.

Comment (WeichenXu123) [reply mlflow/pyfunc/backend.py]: removed the param after pull out "get_or_create_conda_env"

File: mlflow/pyfunc/backend.py
Comment (harupy) [inline mlflow/pyfunc/backend.py:45]: Do we really need to run `python -c ""`?

File: mlflow/utils/conda.py
Comment (harupy) [inline mlflow/utils/conda.py:189]: Can we include the exception in the warning message?

File: mlflow/pyfunc/backend.py
Comment (WeichenXu123) [reply mlflow/pyfunc/backend.py:45]:
```
    if install_mlflow:
        if "MLFLOW_HOME" in os.environ:  # dev version
            install_mlflow = "pip install -e {} 1>&2".format(os.environ["MLFLOW_HOME"])
        else:
            install_mlflow = "pip install mlflow=={} 1>&2".format(VERSION)
```
this part is in `_execute_in_conda_env`, I think 'python -c ""' is for triggering it.

File: mlflow/utils/conda.py
Comment (harupy) [inline mlflow/utils/conda.py:190]:
```suggestion
```

File: mlflow/utils/process.py
Comment (harupy) [inline mlflow/utils/process.py:57]: Are `stdout` and `stderr` binary strings?

File: mlflow/utils/process.py
Comment (harupy) [inline mlflow/utils/process.py:21]:
```suggestion
    capture_output -- if True, captures standard output and error; if False, does not captures these
```

nit: can we swap `True` and `False`?

File: mlflow/utils/process.py
Comment (WeichenXu123) [reply mlflow/utils/process.py:57]: No. because the argument `universal_newlines` is True. which ensures they are str type.

File: mlflow/utils/conda.py
Comment (harupy) [reply mlflow/utils/conda.py:189]: I missed line 186. `capture_output` is set to False. I think we don't need `exc_info=True`.

```suggestion
                _logger.warning(f"Removing conda env '{project_env_name}' failed.")
```

File: mlflow/utils/process.py
Comment (harupy) [reply mlflow/utils/process.py:57]: Got it!

File: mlflow/utils/conda.py
Comment (harupy) [inline mlflow/utils/conda.py:192]:
```suggestion
                _logger.warning(
                    f"Removing conda env '{project_env_name}' failed (error: {repr(e)})."
```

File: mlflow/utils/process.py
Comment (harupy) [inline mlflow/utils/process.py:22]:
```suggestion
    capture_output -- if True, captures standard output and error; if False, does not capture
                      these.
```

File: mlflow/utils/conda.py
Comment (harupy) [inline mlflow/utils/conda.py:104]: This sounds like an exception is not thrown when `capture_output` is False.

File: mlflow/utils/process.py
Comment (harupy) [inline mlflow/utils/process.py:22]:
```suggestion
    capture_output -- if True, captures standard output and error and returns them; if False, does not capture these streams.
```

I think we should indicate standard output and error will be returned.

**Review by harupy — APPROVED** [review]
LGTM once the remaining comments are addressed

File: mlflow/utils/conda.py
Comment (harupy) [inline mlflow/utils/conda.py:106]: To me, this parameter description sounds a bit confusing because `get_or_create_conda_env` doesn't attach stdout/stderr to an exception but `process.exec_cmd` does.

Comment (WeichenXu123) [reply mlflow/utils/conda.py]: updated.

File: mlflow/utils/process.py
Comment (harupy) [inline mlflow/utils/process.py:25]: Why does the parameter description of `capture_output` contain conda?

Comment (WeichenXu123) [reply mlflow/utils/process.py]: removed conda words

**Comment by harupy** [comment]
@dbczumar Could you also take another look since the code has been updated after your approval?

File: mlflow/utils/process.py
Comment (harupy) [inline mlflow/utils/process.py:25]: @WeichenXu123 I think I'd simplify this to the following:

```suggestion
    capture_output -- If True, stdout and stderr will be captured and included in an exception message on failure; if False, these streams won't be captured.
```

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @WeichenXu123 !

**Pull request merged**
